### PR TITLE
swev-id: sympy__sympy-18199 | nthroot_mod: include 0 when a % p == 0 (prime p)

### PR DESCRIPTION
**Pull request by rowan-stein**
Summary

Fix nthroot_mod to include 0 as a solution when p divides a (for prime p). This prevents entering multiplicative-group logic (discrete logs) on s == 0, and returns 0 (or [0] with all_roots=True) as expected.

Linked Issue
- https://github.com/agyn-sandbox/sympy/issues/92

Observed failure and stack trace (before fix)
- Base: branch sympy__sympy-18199
- Repro:

```python
# Ensure mpmath is importable, then:
from sympy.ntheory.residue_ntheory import nthroot_mod
nthroot_mod(11*11, 5, 11)  # expecting 0
```

Stack trace:

```
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
  File "sympy/ntheory/residue_ntheory.py", line 780, in nthroot_mod
    return _nthroot_mod1(a, n, p, all_roots)
  File "sympy/ntheory/residue_ntheory.py", line 728, in _nthroot_mod1
    t = discrete_log(p, s1, h)
  File "sympy/ntheory/residue_ntheory.py", line 1296, in discrete_log
    return _discrete_log_trial_mul(n, a, b, order)
  File "sympy/ntheory/residue_ntheory.py", line 1056, in _discrete_log_trial_mul
    raise ValueError("Log does not exist")
ValueError: Log does not exist
```

Changes
- sympy/ntheory/residue_ntheory.py: short-circuit for a % p == 0 after prime-check; return 0 or [0].
- sympy/ntheory/tests/test_residue.py: add tests covering zero residue cases for several p and n, n==1 identity behavior, and confirm composite modulus remains NotImplemented.

Testing
- Local: ran subset tests
  - PYTHONPATH set to include nix-provided mpmath site-packages.
  - `pytest sympy/ntheory/tests/test_residue.py -q` → 1 passed.
- CI: per request, no CI was triggered.

Notes
- Does not change composite-modulus behavior for nthroot_mod (still NotImplemented).
- This aligns nthroot_mod behavior with sqrt_mod, which already handles a ≡ 0.

### CODE REVIEW

File: sympy/ntheory/residue_ntheory.py [hunk 1]
```diff
@@ -770,11 +770,18 @@ def nthroot_mod(a, n, p, all_roots=False):
...
     if n == 2:
         return sqrt_mod(a, p, all_roots)
+    # For prime moduli, handle zero residue early to include 0 as a root
+    # and avoid incorrect processing in the multiplicative group of units.
+    if not isprime(p):
+        raise NotImplementedError("Not implemented for composite p")
+
+    amod = a % p
+    if amod == 0:
+        return [0] if all_roots else 0
```
Comment (noa-lucent) [inline sympy/ntheory/residue_ntheory.py:787]: [major] The new early return skips the validation that  used to perform. Before this change, inputs with , , or  would raise (or return ) via that helper. Now, when , we return / even for invalid arguments. For example,  and  both silently return , whereas the base branch raises . We need to keep the zero short-circuit but only after the existing validation has run (e.g. call  first, or explicitly replicate those checks) so we don’t weaken the function’s invariants.}

Comment (noa-lucent) [inline sympy/ntheory/residue_ntheory.py:787]: [major] The zero short-circuit now runs before `is_nthpow_residue`, so we no longer trigger the validation that used to enforce `a >= 0`, `n >= 0`, and the special-case handling for `n == 0`. On the base branch, `nthroot_mod(-17, 3, 17)` raises `ValueError('a must be >= 0')` and `nthroot_mod(0, -1, 5)` raises `ValueError('n must be >= 0')`; with this change both calls silently return `0` because `a % p == 0`. Please keep the zero return, but do it only after the existing validation has run (e.g. call `is_nthpow_residue` first, or explicitly replicate those checks) so invalid inputs still surface instead of being accepted as legitimate roots.}

Comment (noa-lucent) [reply sympy/ntheory/residue_ntheory.py:787]: Please ignore this truncated comment; see the follow-up on the same line for the full explanation.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for addressing the missing zero root. While reviewing I noticed the new early return runs before the existing validation logic, so cases like `a < 0` or `n < 0` that used to raise now silently return 0 whenever `a % p == 0`. Please keep the invariants by running the previous validation first (or replicating the checks) before short-circuiting the zero case. Once that is addressed I'm happy to take another look.

**Comment by casey-brooks** [comment]
Tests and lint summary:
- PYTHONPATH=/workspace/pythoncompat:/workspace/sympy PATH=/root/.local/bin:$PATH bin/test sympy/ntheory/tests/test_residue.py → 1 passed, 0 failed
- PYTHONPATH=/workspace/pythoncompat:/workspace/sympy PATH=/root/.local/bin:$PATH bin/test code_quality → 7 passed, 0 failed

File: sympy/ntheory/residue_ntheory.py [hunk 2]
```diff
@@ -770,12 +770,22 @@ def nthroot_mod(a, n, p, all_roots=False):
...
+    residue_exists = is_nthpow_residue(a, n, p)
+
     if not isprime(p):
         raise NotImplementedError("Not implemented for composite p")
 
+    # For prime moduli, handle zero residue early to include 0 as a root
+    # and avoid incorrect processing in the multiplicative group of units.
+    amod = a % p
+    if amod == 0:
+        return [0] if all_roots else 0
```
Comment (noa-lucent) [inline sympy/ntheory/residue_ntheory.py:787]: [major] Thanks for restoring the validation! One more edge case cropped up with the new ordering: when `n == 0` (still invalid input, but previously handled), `is_nthpow_residue(0, 0, p)` returns `False`, so on the base branch `nthroot_mod(0, 0, 5)` returned `None`. With the new code we reach the zero short-circuit before the `residue_exists` guard, so the same call now returns `0`. That’s a behavioral regression. Could you move the `if not residue_exists: return None` check back to before the zero short-circuit (or otherwise ensure the zero path only runs when the residue actually exists)? That way we keep the old semantics for `n == 0` while still fixing the zero root bug.}

**Review by noa-lucent — CHANGES_REQUESTED** [review]
The revision brings back the validation, thanks. There’s still one regression: `nthroot_mod(0, 0, p)` now returns `0` instead of `None` because the zero short-circuit runs before we check `residue_exists`. Please shift that guard (or otherwise gate the zero path) so the old behavior for `n == 0` is preserved.

**Comment by casey-brooks** [comment]
Tests and lint summary:
- PYTHONPATH=/workspace/pythoncompat:/workspace/sympy PATH=/root/.local/bin:$PATH bin/test sympy/ntheory/tests/test_residue.py → 1 passed, 0 failed
- PYTHONPATH=/workspace/pythoncompat:/workspace/sympy PATH=/root/.local/bin:$PATH bin/test code_quality → 7 passed, 0 failed

**Review by noa-lucent — APPROVED** [review]
Thanks for restoring the guard. I re-checked the updated logic and tests: the zero residue short-circuit now runs only after the original validation and residue-existence check, and the new tests cover the regression cases. This matches the spec from issue #92. Looks good to me.
